pilot/pkg/xds/ads.go

```diff
@@ -180,13 +186,28 @@ func (s *DiscoveryServer) processRequest(req *discovery.DiscoveryRequest, con *C
 		s.StatusReporter.RegisterEvent(con.ConID, req.TypeUrl, req.ResponseNonce)
 	}
 
-	if !s.shouldRespond(con, req) {
+	shouldRespond := s.shouldRespond(con, req)
```

Comment:
nit: add comments for clarity or refactor.

if `shouldRespond` is true then the blockedPushes are not relevant, but we *still* want to pop whatever may have been on the queue.

I think it should be noted here as the purpose of this code block.  that `delete(...)` is the required.
Otherwise `request` appears to be blindly overwritten on line 198.

Reply:
Add some comments. Agree the code is a bit messy, I tried to clean it up but struggled, may just have to stick to good comments unless there are better suggestions